### DISCUSSION TITLE
2 Inverters connected to one BMS

### DISCUSSION THREAD

**Discussion by ghost** [Sleeper85/esphome-jk-bms-can#15] [Show and tell]
Today I started to using 2 inverters ( Goodwe )  connected to one BMS ( JKBMS ver 11 )  . I have one ESP board connected using cable to BMS TTL port, and another ESP connected over Bluetooth. Both inverters have their own CAN connected to ESP.  It works for me for few hour without any problem.

**Comment by Sleeper85** [comment]
Perfect, thanks for the info.
I also supervise the same BMS with Wire and BLE.

**Comment by ghost** [comment]
Unfortunatelly, my  BLE version on Olimex LAN doesn't work for me stable . Communication with BMS seem to be correct and stable , but CAN communication works just for 30minutes. Then it says OFF. After downloading ESP again , it works for another 30min. 
Now I am trying to find out what is wrong in my setup. It could be two inverters on one BMS , or Olimex ESP, or becuse I used  3,3V CAN converter . 
For my other two stable connections I am using standart ESP without LAN , wired and with TLA1050 5.0V CAN converter.

**Comment by ghost** [comment]
Hello Sleeper85. Could you please tell me how I can restart ESP32 when Canbus  communication is OFF ?

**Comment by Sleeper85** [comment]
Automatically with programming?

**Reply by ghost** [reply]
Yes. Automatically. Instead of setting message , that it is OFF , just restart it. I am also not at home for now , So  I would like to temporary "solve" my problem until I can figure out , why it is loosing communication after 30 minutes

**Comment by Sleeper85** [comment]
YAML code :
```YAML
button:
  - platform: restart
    name: "Restart button"
    id: restart_button
    internal: true
```

Lambda code :
```CPP
id(restart_button).press();
```


This should be added to this part of the code.
```CPP
              if (id(can_ack_counter) < 20) {                     // Inverter ACK ? => CANBUS ON
                
                id(can_ack_counter)++;                            // CANBUS ACK counter ++
                id(can_msg_counter)++;                            // CANBUS MSG counter ++
                return true;                                      // Condition OK
                
              }
              else if (id(can_status) == "OFF") {                 // CANBUS already OFF ?
                
                return false;                                     // Nothing to do
                
              }
              else {
              
                id(can_status) = "OFF";                           // Set CANBUS Status to OFF
                id(canbus_status).publish_state(id(can_status));  // Publish text sensor
                ESP_LOGI("main", "No rx can 0x305 reply, Inverter not connected/responding...");
                return false;                                     // Condition NOK
                
              }
```

Maybe like this:
```CPP
              if (id(can_ack_counter) < 60) {                     // Inverter ACK ? => CANBUS ON
                
                id(can_ack_counter)++;                            // CANBUS ACK counter ++
                id(can_msg_counter)++;                            // CANBUS MSG counter ++
                return true;                                      // Condition OK
                
              }
              else if (id(can_status) == "OFF") {                 // CANBUS already OFF ?
                
                id(restart_button).press();                       // ESP32 restart
                return false;
                
              }
              else {
              
                id(can_status) = "OFF";                           // Set CANBUS Status to OFF
                id(canbus_status).publish_state(id(can_status));  // Publish text sensor
                ESP_LOGI("main", "No rx can 0x305 reply, Inverter not connected/responding...");
                return false;                                     // Condition NOK
                
              }
```

**Comment by ghost** [comment]
Thank you very much. I found on some forums, to use button for that , but it looks crazy for me. I will try. Thank you very much again
